### PR TITLE
Support installation via containers in new-app

### PR DESCRIPTION
**Pull request by smarterclayton**

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test][extended:core]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@csrwng more review

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -219,13 +233,109 @@ func RunNewApplication(fullName string, f *clientcmd.Factory, out io.Writer, c *
...
+func installationStarted(c kclient.PodInterface, name string, s kclient.SecretsInterface) wait.ConditionFunc {
+	return func() (bool, error) {
+		pod, err := c.Get(name)
+		if err != nil {
+			return false, err
+		}
+		if pod.Status.Phase == kapi.PodPending {
+			return false, nil
+		}
+		// delete a secret named the same as the pod if it exists
```
Comment (csrwng) [inline pkg/cmd/cli/cmd/newapp.go]: I understand why you have to do this now, but it feels ugly... seems like we should allow you to delegate without temporarily creating a secret with your own token. Also, the client may lose the connection while following the installation and then your secret would just hang around.

Comment (liggitt) [inline pkg/cmd/cli/cmd/newapp.go]: > we should allow you to delegate without temporarily creating a secret with your own token

Yes we should, but delegation is a medium-sized feature. Better ideas using the tools we have available now are welcome. In the case of a lost connection, `oc logout` will invalidate the token.

File: pkg/generate/app/cmd/newapp.go [hunk 2]
```diff
@@ -682,23 +732,109 @@ func (c *AppConfig) buildTemplates(components app.ComponentReferences, environme
 	return objects, nil
 }
 
-// ErrNoInputs is returned when no inputs are specified
-var ErrNoInputs = fmt.Errorf("no inputs provided")
+// installComponents locates
```
Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: looks like an unfinished comment

File: pkg/generate/app/cmd/newapp.go [hunk 3]
```diff
@@ -682,23 +732,109 @@ func (c *AppConfig) buildTemplates(components app.ComponentReferences, environme
...
+
+	container, _, err := imageRef.DeployableContainer()
+	if err != nil {
+		return nil, "", fmt.Errorf("can't include %q: %v", input, err)
+	}
+	container.Name = "install"
+
+	// give installers 4 hours to complete
+	deadline := int64(60 * 60 * 4)
+	pod := &kapi.Pod{
```
Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: This code should likely go in an app generator file instead of here.

**Comment by csrwng** [comment]
Overall looks good to me, aside from comments above. I like the splitting out of the imageref stuff. For the installer pod, I think you should follow the same pattern that we use elsewhere to generate API objects, which is a method call on a ref object.

**Comment by smarterclayton** [comment]
What would the ref be?  An imageRef?  Or should it be distinct?

**Comment by csrwng** [comment]
An imageRef works

**Comment by smarterclayton** [comment]
Updated

**Comment by smarterclayton** [comment]
[test]

**Comment by csrwng** [comment]
LGTM

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**